### PR TITLE
Add pre-copy migration support to LXD

### PR DESCRIPTION
**Pull request by adrianreber**
These patches add pre-copy migration support to LXD. These patches are using the existing pre-copy migration support in LXC (LXC needs the following patches https://github.com/lxc/lxc/pull/1950).

Pre-copy migration is based on CRIU and CRIU uses the kernel's dirty memory tracking support:

 https://www.kernel.org/doc/Documentation/vm/soft-dirty.txt
 https://criu.org/Memory_changes_tracking

The pre-copy migration support defaults to off for now as there is no way to query LXC if the architecture/kernel/criu combination support dirty memory tracking. There have been discussion in adding the necessary feature checking to LXC but it is not yet implemented.

If pre-copy migration should be used it is necessary to set 'migration.pre_copy.enabled' to 'true'.

LXD will query if the destination supports pre-copy migration and only then it will start to do multiple pre-copy migration steps. The number of pre-copy iterations can be controlled with 'migration.pre_copy.max' and defaults to 10.

To avoid to do unnecessary pre-copy iterations one more parameter was introduced. With 'migration.pre_copy.pre_migrated_pages' the percentage (defaults to 70%) of pre-copied pages can be controlled if more than 'migration.pre_copy.pre_migrated_pages' have been transferred by the last pre-dump, pre-dumping will stop earlier even if 'migration.pre_copy.max' has not been reached.

This provides the user with two options to control the number of pre-copy migration steps.

### CODE REVIEW

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

File: doc/containers.md [hunk 1]
```diff
@@ -379,3 +379,21 @@ used as a shortcut to set both soft and hard limit (e.g.
...
+equal or smaller than the specified threshold LXD will request CRIU to perform
+a final memory dump and transfer it. If the threshold is not reached after
+a the maximum number of allowed iterations LXD will request a final memory dump
+from CRIU and migrate the container.
+
+Key                                      | Type    | Default | Live update | API extension        | Description
+:--                                      | :---    | :------ | :---------- | :------------        | :----------
+migration.pre\_copy.enabled              | boolean | false   | yes         | migration\_pre\_copy | Use CRIU's pre-copy feature for optimized container memory transfer.
+migration.pre\_copy.max                  | integer | 10      | yes         | migration\_pre\_copy | Maximum number of pre-copy iterations to be performed.
+migration.pre\_copy.pre\_migrated\_pages | integer | 70      | yes         | migration\_pre\_copy | Threshold (%) at which enough memory pages have been pre-copied.
```
Comment (stgraber) [inline doc/containers.md]: All configuration keys should be in the main table at the top.

I'd also recommend the following renames and changes to the descriptions:

 - migration.incremental.memory (was migration.pre_copy.enabled) => Incremental memory transfer of the container's memory to reduce downtime.
 - migration.incremental.memory.goal (was migration.pre_copy.pre_migrated_pages) => Percentage of memory to have in sync before stopping the container.
 - migration.incremental.memory.iterations (was migration.pre_copy.max) => Maximum number of transfer operations to go through before stopping the container.

This should be generic enough to apply to other live migration technologies (should we ever support something other than Linux/LXC/CRIU) and also let us plug in similar incremental transfer for other pieces of the container migration.

I think that also makes it more obvious to a user without CRIU knowledge as to what those knobs actually do.

File: shared/container.go [hunk 2]
```diff
@@ -167,6 +167,10 @@ var KnownContainerConfigKeys = map[string]func(value string) error{
 
 	"linux.kernel_modules": IsAny,
 
+	"migration.pre_copy.enabled":            IsBool,
+	"migration.pre_copy.max":                IsUint32,
+	"migration.pre_copy.pre_migrated_pages": IsUint32,
```
Comment (stgraber) [inline shared/container.go]: The keys should also be listed in the bash completion script.

**Comment by stgraber** [comment]
Other than bikeshedding about key names and missing bash completion, this looks good to me.

File: doc/containers.md (hunk 1 above)
Comment (brauner) [reply doc/containers.md]: changed

File: shared/container.go (hunk 2 above)
Comment (brauner) [reply shared/container.go]: added

**Comment by brauner** [comment]
jenkins: ok to test

**Comment by stgraber** [comment]
@brauner @adrianreber merge conflict

**Comment by brauner** [comment]
@stgraber updated.

**Pull request merged**